### PR TITLE
Documentation for optimised setup

### PR DESCRIPTION
**Pull request by ankkhedia**
*Issue #, if available:*
Issue #350 

*Description of changes:*
The documentation provide information on fine tuning of parameters for performance

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docker/README.md [hunk 1]
```diff
@@ -288,6 +286,10 @@ Notes on a couple of the parameters:
 
     --log-level
     optional
+    
+    ## present only in [mms_app_gpu.conf](./mms_app_gpu.conf)
```
Comment (vdantu) [inline docker/README.md]: This markdown syntax wouldn't work in this section. Probaly change it to" Following option is used only for GPU"

File: docker/README.md [hunk 2]
```diff
@@ -255,21 +255,19 @@ The predict endpoint will return a prediction response in JSON. It will look som
...
 
 Notes on a couple of the parameters:
 
-* **models** - the model used when setting up service. By default it uses resnet-18, chnage this argument to use customized model.
-* **worker-class** - the type of Gunicorn worker processes. We configure by default to gevent which is a type of async worker process. Options are [described in the Gunicorn docs](http://docs.gunicorn.org/en/stable/settings.html#worker-class).
+* **models** - the model used when setting up service. By default it uses resnet-18, change this argument to use customized model.
+* **worker-class** - the type of Gunicorn worker processes. We configure by default to gevent which is a type of async worker process. Options are [described in the Gunicorn docs](http://docs.gunicorn.org/en/stable/settings.html#worker-class). We recommend setting number of workers equal to number of vCPUs in the instance you are using. A detailed discussion of experiments and results can be found [here](../docs/optimised_config.md##number-of-gunicorn-workers(workers))
```
Comment (vdantu) [inline docker/README.md]: This is talking about worker-class. You might want to add a new highlight for "num-workers".

File: docker/README.md [hunk 3]
```diff
@@ -255,21 +255,19 @@ The predict endpoint will return a prediction response in JSON. It will look som
...
 * **limit-request-line** - this is a security-related configuration that limits the [length of the request URI](http://docs.gunicorn.org/en/stable/settings.html#limit-request-line). It is useful preventing DDoS attacks.
-* **num-gpu** - Optional parameter for number of GPUs available for use. MMS uses GPUs with id from 0 .. (num-gpu-1)   By default MMS uses all the available GPUs while this parameter can be configured if user want to use only few of them .
-```
-
+* **num-gpu** - Optional parameter for number of GPUs available for use. MMS uses GPUs with id from 0 .. (num-gpu-1)   By default MMS uses all the available GPUs while this parameter can be configured if user want to use only few of them. A discussion on how to set this parameter can be found [here](../docs/optimised_config.md##number-of-gpus(num-gpu))
```
Comment (vdantu) [inline docker/README.md]: Nit-pick: there is an additional space between (num-gpu-1) and By.

Comment (vdantu) [reply docker/README.md]: And probably a "," after "GPUs and while".

File: docs/optimised_config.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+## Number of Gunicorn workers (workers)
+The number of Gunicorn workers should be equal to the number of vCPUs in the ec2 instance. We varied number of Gunicorn workers and studied throughput and latencies for both GPU(p3.8xlarge) and CPU(c5.2xlarge) instance. The plots below shows how these metrics varies with number of workers.
+
+* Experiments on GPU( p3.8xlarge with 32 vCPUs and 4 GPUs)
+The plots below shows how throughput and median latency varied for 100 requests from 100 concurrent workers with number of Gunicorn workers on p3.8xlarge instance. 
+![GPU_throughput](docs/images/gpu_throughput.png)
+![GPU_latency](docs/images/gpu_latency.png)
+
+We can see that we the throughput becomes constant after 32 workers while latency increases a bit suggesting 32 workers to be optimal for the instance.
```
Comment (vdantu) [inline docs/optimised_config.md]: remove the second "we"

File: docs/optimised_config.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+## Number of GPUs (num-gpu)
+The best performances are obtained using all the available GPUs available on the system. Experiments shows that it linearly scales throughput . By default, MMS identifies number of available GPUs and assign context of Gunicorn worker threads to each of them in round robin fashion. However, you can configure the number of GPUs in  you want to use in the [mms_app_gpu.conf](../docker/mms_app_gpu.conf) to use only few of the available instances.
+
+## Performance on high loads 
+After setting number of workers and GPUs as described above, we ran experiments to understand scale of request which MMS can handle. The containerised GPU version of MMS was able to give throughput of 650 requests/second without any error when it was bombarded by request from 600 concurrent workers sending 100 request each. 
+
```
Comment (vdantu) [inline docs/optimised_config.md]: Remove extra empty lines

File: docs/optimised_config.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
+# Optimised configuration for EC2 instances for mms in containers
```
Comment (vdantu) [inline docs/optimised_config.md]: Captilize mms

**Comment by lupesko** [comment]
@aaronmarkham can you take a look?
@ankkhedia - can we please link this new page from https://github.com/awslabs/mxnet-model-server/tree/master/docs root readme ?

**Comment by lupesko** [comment]
@ankkhedia are you sure the CPU latency numbers are correct? 761ms for inference on resnet-18 is very high!

**Comment by ankkhedia** [comment]
@lupesko  The CPU numbers are correct. I reran the experiments for Cpu and found similar numbers. 


The higher latencies is attributed to higher number of requests coming in. We ran 100 concurrent workers sending 100 request each to 32 vCPU in P3.8xlarge. In case of CPU(c5.2xlarge) , the number of vCPUs is only 8 whereas the request rate remains same and hence we got the backlog and higher latencies. Apart from that some of the work gets shared by GPU in p3.8xlarge leading to better latencies. 

We tried the same experiments with 25 concurrent workers sending 100 request each to the CPU instance and the latencies were comparable to that for 10000 requests for GPU instance. So, it is basically the number of backlog requests leading to high latencies

**Comment by lupesko** [comment]
Thanks @ankkhedia 
It may be better to re-run and update the docs with lower concurrency since 761ms for resnet-18 is really too high...
Let me know what you think.

**Pull request merged**